### PR TITLE
Make vddk log level configurable

### PR DESCRIPTION
**Pull request by xinyanw409**
Currently we hardcoding the default log level by explicitly calling VixDiskLib_Init in VDDK side. 

VDDK has a function VixDiskLib_InitEx which allows user to pass a config file which contains customized log level configuration. We have already added the gvddk wrapper function for VixDiskLib_InitEx, we need to update our vsphere plugin to allow user to configure log level during run time.

This change includes:
1. Pick up virtual-disks change.
2. Change to use disklib.InitEx and pass a config file path when initializing vddk lib.

Manually tested. Detail test steps refer to https://github.com/vmware-tanzu/velero-plugin-for-vsphere/pull/354.

### CODE REVIEW

File: pkg/ivd/ivd_protected_entity_type_manager.go [hunk 1]
```diff
@@ -378,11 +379,27 @@ func (this *IVDProtectedEntityTypeManager) ReloadConfig(ctx context.Context, par
...
+		err = disklib.Init(vsphereMajor, vSphereMinor, disklibLib64)
+		if err != nil {
+			return errors.Wrap(err, "Could not initialize VDDK during config reload.")
+		}
+	} else {
+		path, err := util.CreateConfigFile(params, this.logger)
+		if err != nil {
+			this.logger.Errorf("Failed to create config file for vddk. Cannot proceed to init vddk lib")
+			return errors.Wrap(err, "Failed to create config file")
+		}
```
Comment (xinyanw409) [inline pkg/ivd/ivd_protected_entity_type_manager.go:392]: path-> /tmp

File: pkg/ivd/ivd_protected_entity_type_manager.go [hunk 2]
```diff
@@ -378,11 +379,27 @@ func (this *IVDProtectedEntityTypeManager) ReloadConfig(ctx context.Context, par
 		this.cnsManager.ResetManager(reloadedVc, cnsClient)
 	}
 
-	err = disklib.Init(vsphereMajor, vSphereMinor, disklibLib64)
-	if err != nil {
-		return errors.Wrap(err, "Could not initialize VDDK during config reload.")
+	if _, ok := params["vddkLogLevel"]; !ok {
+		err = disklib.Init(vsphereMajor, vSphereMinor, disklibLib64)
```
Comment (lintongj) [inline pkg/ivd/ivd_protected_entity_type_manager.go:383]: Do you still plan to stick with the deprecating VDDK API?

File: pkg/ivd/ivd_protected_entity_type_manager.go [hunk 3]
```diff
@@ -378,11 +379,27 @@ func (this *IVDProtectedEntityTypeManager) ReloadConfig(ctx context.Context, par
...
+		if err != nil {
+			return errors.Wrap(err, "Could not initialize VDDK during config reload.")
+		}
+	} else {
+		path, err := util.CreateConfigFile(params, this.logger)
+		if err != nil {
+			this.logger.Errorf("Failed to create config file for vddk. Cannot proceed to init vddk lib")
```
Comment (lintongj) [inline pkg/ivd/ivd_protected_entity_type_manager.go:390]: Errorf -> Error

File: pkg/util/utils.go [hunk 4]
```diff
@@ -181,3 +182,65 @@ func GetClusterFlavor(config *rest.Config) (ClusterFlavor, error) {
 	// Did not match any search criteria. Unknown cluster flavor.
 	return Unknown, errors.New("GetClusterFlavor: Failed to identify cluster flavor")
 }
+
+func CreateConfigFile(params map[string]interface{}, logger logrus.FieldLogger) (string, error) {
+	path := "/tmp/config"
+	logLevel := params["vddkLogLevel"].(string)
+	err := createFile(path, logLevel, logger)
+	if err != nil {
+		logger.WithError(err).Errorf("Failed to create config file")
```
Comment (lintongj) [inline pkg/util/utils.go:191]: Errorf -> Error

File: pkg/util/utils.go [hunk 5]
```diff
@@ -181,3 +182,65 @@ func GetClusterFlavor(config *rest.Config) (ClusterFlavor, error) {
...
+		return err
+	}
+
+	// Save file changes.
+	err = file.Sync()
+	if err != nil {
+		return err
+	}
+
+	logger.Infof("Config File Created Successfully", path)
```
Comment (lintongj) [inline pkg/util/utils.go:232]: You might want add a `%v` or `%s` for the path.

File: pkg/ivd/ivd_protected_entity_type_manager.go [hunk 6]
```diff
@@ -378,11 +380,32 @@ func (this *IVDProtectedEntityTypeManager) ReloadConfig(ctx context.Context, par
...
-	if err != nil {
-		return errors.Wrap(err, "Could not initialize VDDK during config reload.")
+	// check whether customized vddk config is provided.
+	// currently only support user to modify vixdisklib nfc log level and transport log level
+	if _, ok := params[vddkconfig]; !ok {
+		this.logger.Info("No customized vddk log level provided, set vddk log level as default.")
+		err = disklib.InitEx(vsphereMajor, vSphereMinor, disklibLib64, "")
+		if err != nil {
+			return errors.Wrap(err, "Could not initialize VDDK during config reload.")
+		}
```
Comment (lintongj) [inline pkg/ivd/ivd_protected_entity_type_manager.go:390]: the code snippet L387-L390 is used twice. Would you please re-organize your code to get rid of the repeated code snippet?

File: pkg/ivd/ivd_protected_entity_type_manager.go [hunk 7]
```diff
@@ -378,11 +380,32 @@ func (this *IVDProtectedEntityTypeManager) ReloadConfig(ctx context.Context, par
 		this.cnsManager.ResetManager(reloadedVc, cnsClient)
 	}
 
-	err = disklib.Init(vsphereMajor, vSphereMinor, disklibLib64)
-	if err != nil {
-		return errors.Wrap(err, "Could not initialize VDDK during config reload.")
+	// check whether customized vddk config is provided.
+	// currently only support user to modify vixdisklib nfc log level and transport log level
+	if _, ok := params[vddkconfig]; !ok {
+		this.logger.Info("No customized vddk log level provided, set vddk log level as default.")
```
Comment (lintongj) [inline pkg/ivd/ivd_protected_entity_type_manager.go:386]: Please double check all logging statement in your PR and remove the end period in each of them.

>this.logger.Info("No customized vddk log level provided, set vddk log level as default")

File: pkg/util/utils.go [hunk 8]
```diff
@@ -181,3 +182,68 @@ func GetClusterFlavor(config *rest.Config) (ClusterFlavor, error) {
...
+		}
+	}
+
+	// Save file changes.
+	err = file.Sync()
+	if err != nil {
+		return err
+	}
+
+	logger.Infof("Config File %v Created Successfully", path)
```
Comment (lintongj) [inline pkg/util/utils.go:235]: Infof -> Debugf?

File: pkg/util/utils.go [hunk 9]
```diff
@@ -181,3 +182,68 @@ func GetClusterFlavor(config *rest.Config) (ClusterFlavor, error) {
 	// Did not match any search criteria. Unknown cluster flavor.
 	return Unknown, errors.New("GetClusterFlavor: Failed to identify cluster flavor")
 }
+
+func CreateConfigFile(vddkConfig map[string]string, logger logrus.FieldLogger) (string, error) {
+	path := "/tmp/config"
```
Comment (lintongj) [inline pkg/util/utils.go:187]: Please do not hardcode anything in func. Make a constant and use it here.

File: pkg/ivd/ivd_protected_entity_type_manager.go [hunk 10]
```diff
@@ -378,11 +380,32 @@ func (this *IVDProtectedEntityTypeManager) ReloadConfig(ctx context.Context, par
 		this.cnsManager.ResetManager(reloadedVc, cnsClient)
 	}
 
-	err = disklib.Init(vsphereMajor, vSphereMinor, disklibLib64)
+	// check whether customized vddk config is provided.
+	// currently only support user to modify vixdisklib nfc log level and transport log level
+	var path string
+	if _, ok := params[vddkconfig]; !ok {
+		this.logger.Info("No customized vddk log level provided, set vddk log level as default")
+		path = ""
```
Comment (lintongj) [inline pkg/ivd/ivd_protected_entity_type_manager.go:388]: this line is unnecessary

File: pkg/ivd/ivd_protected_entity_type_manager.go [hunk 11]
```diff
@@ -378,11 +380,32 @@ func (this *IVDProtectedEntityTypeManager) ReloadConfig(ctx context.Context, par
...
+	} else {
+		this.logger.Info("Customized vddk config provided: %v", params[vddkconfig])
```
Comment (lintongj) [inline pkg/ivd/ivd_protected_entity_type_manager.go:390]: Infof

File: pkg/util/utils.go [hunk 12]
```diff
@@ -35,6 +36,8 @@ const (
 	VSphere                  = "vSphere Kubernetes Cluster"
 )
 
+const VddkConfigPath = "/tmp/config"
```
Comment (lintongj) [inline pkg/util/utils.go:39]: Would you please move this constant definition to the top of this file? Eventually, we might want to move them to a separate file.

**Review by lintongj — APPROVED** [review]
Looks good to me except some minor comments

**Pull request merged**